Bug when running mypy on a sigle file (with per-file settings in the config)

The use case: I don't want to type-check tests and other misc files. I'm using vim with the syntastic plugin, which runs mypy on a file affter saving it.

This is my mypy config
```
$ cat mypy.ini 
[mypy]
silent_imports = True

[mypy-module/*]
silent_imports = False
```

This is my code
```
$ cat module/base.py 
class Config:
    pass
```
```
$ cat module/feature.py 
from module.base import Config


def do_things() -> None:
    use_config(False)  # error


def use_config(config: Config) -> None:
    return
```

Running mypy on the module directory works as expected
```
$ mypy module
module/feature.py: note: In function "do_things":
module/feature.py:5: error: Argument 1 to "use_config" has incompatible type "bool"; expected "Config"
```

Running mypy on a single file (that's how syntastic uses mypy) reports no errors
```
$ mypy module/feature.py
(no errors)
```
```

--- a/docs/source/config_file.rst
+++ b/docs/source/config_file.rst
@@ -10,7 +10,7 @@ file instead (see :ref:`--config-file <config-file-flag>`).
 
 Most flags correspond closely to :ref:`command-line flags
 <command-line>` but there are some differences in flag names and some
-flags may take a different value based on the file being processed.
+flags may take a different value based on the module being processed.
 
 The configuration file format is the usual
 `ini file <https://docs.python.org/3.6/library/configparser.html>`_
@@ -22,10 +22,10 @@ characters.
   the global flags.
 
 - Additional sections named ``[mypy-PATTERN1,PATTERN2,...]`` may be
-  present, where ``PATTERN1``, ``PATTERN2`` etc. are `filename
-  patterns <https://docs.python.org/3.6/library/fnmatch.html>`_
+  present, where ``PATTERN1``, ``PATTERN2`` etc. are `fnmatch patterns
+  <https://docs.python.org/3.6/library/fnmatch.html>`_
   separated by commas.  These sections specify additional flags that
-  only apply to files whose name matches at least one of the patterns.
+  only apply to *modules* whose name matches at least one of the patterns.
 
 Global flags
 ************
@@ -98,16 +98,16 @@ The following global flags may only be set in the global section
   error messages.
 
 
-Per-file flags
-**************
+Per-module flags
+****************
 
-The following flags may vary per file.  They may also be specified in
+The following flags may vary per module.  They may also be specified in
 the global section; the global section provides defaults which are
-overridden by the pattern sections matching the file.
+overridden by the pattern sections matching the module name.
 
 .. note::
 
-   If multiple pattern sections match a file they are processed in
+   If multiple pattern sections match a module they are processed in
    unspecified order.
 
 - ``silent_imports`` (Boolean, default False) silences complaints
@@ -144,13 +144,13 @@ You might put this in your ``mypy.ini`` file at the root of your repo:
 
     [mypy]
     python_version = 2.7
-    [mypy-foo/*]
+    [mypy-foo.*]
     disallow_untyped_defs = True
 
 This automatically sets ``--python-version 2.7`` (a.k.a. ``--py2``)
 for all mypy runs in this tree, and also selectively turns on the
-``--disallow-untyped-defs`` flag for all files under the ``foo/``
-subdirectory.  This issues an error for function definitions without
+``--disallow-untyped-defs`` flag for all modules in the ``foo``
+package.  This issues an error for function definitions without
 type annotations in that subdirectory only.
 
 .. note::

--- a/mypy/build.py
+++ b/mypy/build.py
@@ -879,7 +879,7 @@ def write_cache(id: str, path: str, tree: MypyFile,
     st = manager.get_stat(path)  # TODO: Handle errors
     mtime = st.st_mtime
     size = st.st_size
-    options = manager.options.clone_for_file(path)
+    options = manager.options.clone_for_module(id)
     meta = {'id': id,
             'path': path,
             'mtime': mtime,
@@ -1119,7 +1119,7 @@ class State:
         else:
             self.import_context = []
         self.id = id or '__main__'
-        self.options = manager.options.clone_for_file(path or '')
+        self.options = manager.options.clone_for_module(self.id)
         if not path and source is None:
             file_id = id
             if id == 'builtins' and self.options.python_version[0] == 2:
@@ -1508,7 +1508,7 @@ def dispatch(sources: List[BuildSource], manager: BuildManager) -> None:
     manager.log("Loaded graph with %d nodes" % len(graph))
     process_graph(graph, manager)
     if manager.options.warn_unused_ignores:
-        # TODO: This could also be a per-file option.
+        # TODO: This could also be a per-module option.
         manager.errors.generate_unused_ignore_notes()
 
 

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -484,19 +484,22 @@ def parse_config_file(options: Options, filename: str) -> None:
         if name.startswith('mypy-'):
             prefix = '%s: [%s]' % (filename, name)
             updates, report_dirs = parse_section(prefix, options, section)
-            # TODO: Limit updates to flags that can be per-file.
             if report_dirs:
-                print("%s: Per-file sections should not specify reports (%s)" %
+                print("%s: Per-module sections should not specify reports (%s)" %
                       (prefix, ', '.join(s + '_report' for s in sorted(report_dirs))),
                       file=sys.stderr)
-            if set(updates) - Options.PER_FILE_OPTIONS:
-                print("%s: Per-file sections should only specify per-file flags (%s)" %
-                      (prefix, ', '.join(sorted(set(updates) - Options.PER_FILE_OPTIONS))),
+            if set(updates) - Options.PER_MODULE_OPTIONS:
+                print("%s: Per-module sections should only specify per-module flags (%s)" %
+                      (prefix, ', '.join(sorted(set(updates) - Options.PER_MODULE_OPTIONS))),
                       file=sys.stderr)
-                updates = {k: v for k, v in updates.items() if k in Options.PER_FILE_OPTIONS}
+                updates = {k: v for k, v in updates.items() if k in Options.PER_MODULE_OPTIONS}
             globs = name[5:]
             for glob in globs.split(','):
-                options.per_file_options[glob] = updates
+                # For backwards compatibility, replace (back)slashes with dots.
+                glob = glob.replace(os.sep, '.')
+                if os.altsep:
+                    glob = glob.replace(os.altsep, '.')
+                options.per_module_options[glob] = updates
 
 
 def parse_section(prefix: str, template: Options,

--- a/mypy/options.py
+++ b/mypy/options.py
@@ -121,7 +121,7 @@ class Options:
     def clone_for_module(self, module: str) -> 'Options':
         updates = {}
         for pattern in self.per_module_options:
-            if self.module_matches_pattern(pattern):
+            if self.module_matches_pattern(module, pattern):
                 updates.update(self.per_module_options[pattern])
         if not updates:
             return self

--- a/mypy/options.py
+++ b/mypy/options.py
@@ -16,7 +16,7 @@ class BuildType:
 class Options:
     """Options collected from flags."""
 
-    PER_FILE_OPTIONS = {
+    PER_MODULE_OPTIONS = {
         "silent_imports",
         "almost_silent",
         "disallow_untyped_calls",
@@ -27,7 +27,7 @@ class Options:
         "show_none_errors",
     }
 
-    OPTIONS_AFFECTING_CACHE = PER_FILE_OPTIONS | {"strict_optional"}
+    OPTIONS_AFFECTING_CACHE = PER_MODULE_OPTIONS | {"strict_optional"}
 
     def __init__(self) -> None:
         # -- build options --
@@ -83,8 +83,8 @@ class Options:
         # Write junit.xml to given file
         self.junit_xml = None  # type: Optional[str]
 
-        # Per-file options (raw)
-        self.per_file_options = {}  # type: Dict[str, Dict[str, object]]
+        # Per-module options (raw)
+        self.per_module_options = {}  # type: Dict[str, Dict[str, object]]
 
         # -- development options --
         self.verbosity = 0  # More verbose messages (for troubleshooting)
@@ -118,17 +118,23 @@ class Options:
     def __repr__(self) -> str:
         return 'Options({})'.format(pprint.pformat(self.__dict__))
 
-    def clone_for_file(self, filename: str) -> 'Options':
+    def clone_for_module(self, module: str) -> 'Options':
         updates = {}
-        for glob in self.per_file_options:
-            if fnmatch.fnmatch(filename, glob):
-                updates.update(self.per_file_options[glob])
+        for pattern in self.per_module_options:
+            if self.module_matches_pattern(pattern):
+                updates.update(self.per_module_options[pattern])
         if not updates:
             return self
         new_options = Options()
         new_options.__dict__.update(self.__dict__)
         new_options.__dict__.update(updates)
         return new_options
+
+    def module_matches_pattern(self, module: str, pattern: str) -> bool:
+        # If the pattern is 'mod.*', we want 'mod' to match that too.
+        # (That's so that a pattern specifying a package also matches
+        # that package's __init__.)
+        return fnmatch.fnmatch(module, pattern) or fnmatch.fnmatch(module + '.', pattern)
 
     def select_options_affecting_cache(self) -> Mapping[str, bool]:
         return {opt: getattr(self, opt) for opt in self.OPTIONS_AFFECTING_CACHE}

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -177,7 +177,7 @@ class SemanticAnalyzer(NodeVisitor):
     bound_tvars = None  # type: List[SymbolTableNode]
     # Stack of type variables that were bound by outer classess
     tvar_stack = None  # type: List[List[SymbolTableNode]]
-    # Per-file options
+    # Per-module options
     options = None  # type: Options
 
     # Stack of functions being analyzed

--- a/mypy_strict_optional.ini
+++ b/mypy_strict_optional.ini
@@ -4,5 +4,5 @@
 strict_optional = True
 show_none_errors = False
 
-[mypy-mypy/test/*]
+[mypy-mypy.test.*]
 show_none_errors = True

--- a/test-data/unit/cmdline.test
+++ b/test-data/unit/cmdline.test
@@ -162,9 +162,9 @@ x.py:1: error: Function is missing a type annotation
 [file mypy.ini]
 [[mypy]
 hide_error_context = True
-[[mypy-*x*.py]
+[[mypy-*x*]
 disallow_untyped_defs = True
-[[mypy-*y*.py]
+[[mypy-*y*]
 disallow_untyped_calls = True
 [file xx.py]
 def f(a): pass
@@ -191,7 +191,7 @@ xx.py:1: error: Function is missing a type annotation
 [file mypy.ini]
 [[mypy]
 hide_error_context = True
-[[mypy-x*py,z*py]
+[[mypy-x*,z*]
 disallow_untyped_defs = True
 [file x.py]
 def f(a): pass

--- a/test-data/unit/cmdline.test
+++ b/test-data/unit/cmdline.test
@@ -240,7 +240,7 @@ mypy.ini: [mypy]: silent_imports: Not a boolean: nah
 [[mypy-*]
 strict_optional = True
 [out]
-mypy.ini: [mypy-*]: Per-file sections should only specify per-file flags (strict_optional)
+mypy.ini: [mypy-*]: Per-module sections should only specify per-module flags (strict_optional)
 
 [case testCoberturaParser]
 # cmd: mypy --cobertura-xml-report build pkg

